### ISSUE TITLE
WAMP interface definition and payload validation

### ISSUE THREAD

**Issue by oberstet** [crossbario/crossbar#302]
The routing core of Crossbar.io already has an internal hook for validating the application payload of WAMP calls or events [here](https://github.com/crossbario/crossbar/blob/master/crossbar/router/router.py#L386).

This hook is planned to be used for one of the use cases of WAMP interface definitions: payload validation. (For other use cases, see below).

> Note: this stuff has been discussed in various contexts over the past, both here in issues, on Autobahn and WAMP repos and mailing lists. You have to search ..

**Comment by oberstet** [comment]
#301, #302 and #303 are all related

**Comment by oberstet** [comment]
Essentially, we have discussed this topic in the past in various contexts:

1. WAMP interface definitions (using an IDL)
2. WAMP interface run-time validation
3. WAMP interface run-time reflection
4. WAMP interface documentation generation
5. WAMP proxy/stub generation (for statically compiled languages)

Here, 2. - 5. are the actual use cases, technically building on 1.

Use cases 2. + 3. are naturally driven from Crossbar.io itself: to validate the payload of WAMP Calls and Events at run-time, the natural place to centrally and strictly enforce this is at the router level. Same with 3.: a meta API allowing client programs to access WAMP interface definitions at run-time has to be implemented inside Crossbar.io (in the service session).

For use case 4., the natural way in which we can add this is via a new Web path resource: a Jinja2 template driven Twisted Web resource that can be configured into a Web transport of Crossbar.io, and which renders the WAMP interface definitions loaded inside Crossbar.io as HTML pages. These pages are intended to be accessed by application developers.

Lastly, use case 5.: this needs to be done externally by a code generator tool. That tool could access the WAMP interface definitions via the Web resource used for documentation, but request a content return type of JSON instead of HTML.

---

**To summarize, all use cases we want to cover can be naturally addressed in Crossbar.io by adding the ability to load WAMP interface definitions into router worker (realms).**

**Comment by oberstet** [comment]
Here is an example node configuration that shows how to specify WAMP interface definitions:

```json
{
    "workers": [
        {
            "type": "router",
            "realms": [
                {
                    "name": "realm1",
                    "roles": [
                    ],
                    "interfaces": [
                        {
                            "id": "interface-001",
                            "uri": "com.symanto.fanalytics.fan.<fan_id:int>.",
                            "match": "prefix",

                            "title": "Fanalytics Fan Profile API",
                            "description": "The Fanalytics provides AI powered sports fan profiles and can be accessed via this API.",

                            "definitions": {
                                "fan_profile": {
                                    "type": "object",
                                    "properties": {
                                        "fan_id": {
                                            "type": "integer"
                                        },
                                        "last_update": {
                                            "type": "string"
                                        },
                                        "fan_persona": {
                                            "type": "string"
                                        },
                                        "favorite_club": {
                                            "type": "string"
                                        },
                                        "club_affinity_score": {
                                            "type": "string"
                                        },
                                    },
                                    "required": ["fan_id", "last_update"]
                                }
                            },

                            "elements": {

                                "get_profile@1": {
                                    "type": "procedure",
                                    "parameters": {
                                        "args": [
                                            {
                                                "type": "integer",
                                                "minimum": 7,
                                                "title": "Average over Days",
                                                "description": "Return profile averaged over this many days."
                                            }
                                        ],
                                        "kwargs": {
                                            "include_scores": {
                                                "type": "boolean",
                                                "title": "Include Scores",
                                                "description": "Flag to request including score values in returned fan profile."
                                            }
                                        }
                                    },
                                    "progress": null,
                                    "result": {
                                        "args": [
                                            {
                                                "$ref": "#/definitions/fan_profile"
                                            }
                                        ],
                                        "kwargs": null
                                    },
                                    "error": null
                                },

                                "on_profile_changed@1": {
                                    "type": "event",
                                    "payload": {
                                        "args": [
                                            {
                                                "$ref": "#/definitions/fan_profile"
                                            }
                                        ],
                                        "kwargs": null
                                    }
                                }

                            }
                        }
                    ]
                }
            ]
        }
    ]
}
```

**Comment by oberstet** [comment]
To implement above, we need 2 pieces:

1. [ ] extension of node configuration checking and parsing
2. [ ] new management procedures inside the router worker

The latter requires 3 new management procedures in [here](https://github.com/crossbario/crossbar/blob/master/crossbar/worker/router.py) (plus respective management events):

* `get_router_realm_interfaces`
* `start_router_realm_interface`
* `stop_router_realm_interface`

**Comment by oberstet** [comment]
After above is there, we can then work on

```json
                            "enable": {
                                "validation": true,
                                "docs": true,
                                "reflection": true
                            },
```

That is, above flags enable the use of the interface definitions for different purposes.

Eg the validation would need to hook into the payload validation inside the router and check that the application URIs used are used are allowed for the respective use (eg as a Call URI), and that the payloads conform to the JSON Schema snippets defined for the interfaces.

**Comment by oberstet** [comment]
@om26er not sure if above is sufficient info to understand the idea, but this would be a nice first feature you could work on within the Crossbar.io code base. Because it is new, and because it has very well defined internal interfaces (the places where to add what are clear). Pls have a look and let me know if you feel like working on this;)

**Comment by om26er** [comment]
> @om26er not sure if above is sufficient info to understand the idea, but this would be a nice first feature you could work on within the Crossbar.io code base. Because it is new, and because it has very well defined internal interfaces (the places where to add what are clear). Pls have a look and let me know if you feel like working on this;)

With the discussion we had yesterday and rereading have made things much cleared to me. We will definitely need to refine the actual interface but I now have everything to get started.

**Comment by oberstet** [comment]
Issue migrated to Crossbar.io Fabric repo
